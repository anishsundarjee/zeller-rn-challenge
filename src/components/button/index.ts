export * from './customButton';
export * from './backButton';