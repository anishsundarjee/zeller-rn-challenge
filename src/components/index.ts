export * from './container';
export * from './spacing';
export * from './commonScreens';
export * from './button';
export * from './input';