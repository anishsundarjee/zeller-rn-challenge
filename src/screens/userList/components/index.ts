export * from './userTypeSelector';
export * from './user';
export * from './userFilterInput';