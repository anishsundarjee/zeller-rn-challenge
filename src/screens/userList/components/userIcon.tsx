import { COLORS } from '@/constants';
import React from 'react';
import { StyleSheet } from 'react-native';
import { View, Text } from 'react-native-ui-lib';

type UserIconProps = {
  testID: string;
  initial: string;
};

export const UserIcon = ({ initial, testID }: UserIconProps) => {
  return (
    <View testID={testID} style={styles.container}>
      <Text blue body700>{initial}</Text>
    </View>
  )
};

const styles = StyleSheet.create({
  container: {
    width: 50,
    height: 50,
    borderRadius: 8,
    backgroundColor: COLORS.lightBlue,
    justifyContent: 'center',
    alignItems: 'center',
  }
})