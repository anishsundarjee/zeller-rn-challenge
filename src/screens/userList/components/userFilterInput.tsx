import React, { useState } from 'react';
import { TextInput } from '@/components';
import { STRINGS } from '@/constants';
import { TFilterOptions } from '@/types';

type UserFilterInputProps = {
  handleFilterChange: (value: TFilterOptions) => void;
};

export const UserFilterInput = ({ handleFilterChange }: UserFilterInputProps) => {
  const [value, setValue] = useState('');

  const handleValueChange = (value: string) => {
    setValue(value);
    const filterOption: TFilterOptions = {
      label: 'Search',
      value: value,
      type: 'name',
      isSelected: true,
    };

    handleFilterChange(filterOption);
  };

  return (
    <TextInput
      testID='user-filter-input'
      value={value}
      placeholder={STRINGS.userList.searchUsers}
      label={STRINGS.userList.searchUsers}
      onChangeText={handleValueChange}
    />
  );
};