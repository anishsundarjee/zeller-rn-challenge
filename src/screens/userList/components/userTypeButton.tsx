import React from 'react';
import { StyleSheet } from 'react-native';
import { RadioButton, Text, TouchableOpacity } from 'react-native-ui-lib';
import { TFilterOptions } from '@/types';
import { COLORS } from '@/constants';
import { Padder } from '@/components';

type UserTypeButtonProps = {
  testID?: string;
  option: TFilterOptions;
  handleFilterChange: (value: TFilterOptions) => void;
};

export const UserTypeButton = ({ option, handleFilterChange, testID = 'user-type-btn' }: UserTypeButtonProps) => {
  const backGroundColor = option.isSelected ? COLORS.lightBlue : COLORS.white;
  return (
    <TouchableOpacity
      testID={testID}
      onPress={() => handleFilterChange(option)}
      style={[styles.button, { backgroundColor: backGroundColor }]}
      activeOpacity={0.7}
      disabled={option.isSelected}
    >
      <RadioButton
        testID={'radio-btn'}
        value={undefined}
        selected={option.isSelected}
        size={20}
        color={COLORS.blue}
      />
      <Padder />
      <Text body black testID={`user-type-txt-${option.label}`}>{option.label}</Text>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  button: {
    paddingVertical: 16,
    paddingHorizontal: 16,
    marginBottom: 4,
    borderRadius: 8,
    flexDirection: 'row',
    alignItems: 'center',
  }
})