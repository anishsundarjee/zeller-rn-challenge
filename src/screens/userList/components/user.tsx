import React from 'react';
import { StyleSheet } from 'react-native';
import { View, Text, TouchableOpacity } from 'react-native-ui-lib';
import { TCustomer } from '@/types';
import { UserIcon } from './userIcon';
import { COLORS } from '@/constants';
import { Padder } from '@/components';

type UserProps = {
  item: TCustomer;
  onPress: (item: TCustomer) => void;
};

export const User = ({ item, onPress }: UserProps) => {
  const { name, role } = item;
  return (
    <TouchableOpacity
      testID={`user-btn-${name}`}
      onPress={() => onPress(item)}
      style={styles.button}
      activeOpacity={0.7}
    >
      <UserIcon testID={`user-icon-${name}`} initial={name[0]} />
      <Padder />
      <View>
        <Text body700 black>{name}</Text>
        <Text body black>{role}</Text>
      </View>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  button: {
    paddingBottom: 8,
    backgroundColor: COLORS.white,
    marginBottom: 4,
    flexDirection: 'row',
    alignItems: 'center',
  }
});