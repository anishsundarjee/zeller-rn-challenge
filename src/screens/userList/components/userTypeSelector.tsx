import React from 'react';
import { View, Text } from 'react-native-ui-lib';
import { Padder } from '@/components';
import { STRINGS, COLORS } from '@/constants';
import { TFilterOptions } from '@/types';
import { UserTypeButton } from './userTypeButton';

type UserTypeSelectorProps = {
  filterOptions: TFilterOptions[];
  handleFilterChange: (value: TFilterOptions) => void;
};

export const UserTypeSelector = ({ filterOptions, handleFilterChange }: UserTypeSelectorProps) => {
  const selectedOption = filterOptions.find((option) => option.isSelected);
  return (
    <View>
      {filterOptions.map((option) => (
        <UserTypeButton
          testID={`user-type-btn-${option.label}`}
          key={option.value}
          option={option}
          handleFilterChange={handleFilterChange}
        />
      ))}
      <Padder h={24} />
      <Padder h={1} bg={COLORS.grey} wide={true} />
      <Padder h={24} />
      <Text h5>
        {`${selectedOption ? selectedOption.label : 'All'} ${STRINGS.userList.users}`}
      </Text>
      <Padder h={16} />
    </View>
  );
};