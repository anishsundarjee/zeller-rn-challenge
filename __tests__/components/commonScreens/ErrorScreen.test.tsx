import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import { it, describe, expect } from '@jest/globals';
import { ErrorScreen } from '@/components';

jest.mock('react-native-safe-area-context', () => {
  return {
    useSafeAreaInsets: () => ({ top: 0, bottom: 0 }),
  };
});

describe('ErrorScreen', () => {
  it('compare snapshot', async () => {
    const tree = renderer.create(
      <ErrorScreen onPress={jest.fn()} testID='error-screen' />,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
  it('renders correctly and press button', async () => {
    const mockPress = jest.fn();
    const tree = renderer.create(
      <ErrorScreen onPress={mockPress} testID='error-screen'/>,
    );

    expect(tree.root.findByProps({ testID: 'error-screen' }).props).toBeDefined();
    const button = tree.root.findByProps({ testID: 'error-button' }).props;
    expect(button).toBeDefined();

    await renderer.act(async () => {
      await button.onPress();
    });

    expect(mockPress).toHaveBeenCalled();
  });
});