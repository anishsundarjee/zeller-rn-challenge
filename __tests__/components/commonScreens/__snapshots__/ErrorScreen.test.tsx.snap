// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorScreen compare snapshot 1`] = `
<View
  style={
    [
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      [
        {
          "backgroundColor": "#FFFFFF",
          "flex": 1,
          "justifyContent": "space-between",
        },
        {
          "paddingTop": 0,
        },
        {
          "alignItems": "center",
          "justifyContent": "center",
        },
      ],
    ]
  }
  testID="error-screen"
>
  <View
    style={
      [
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        [
          {
            "paddingBottom": 24,
            "paddingTop": 16,
          },
          {
            "paddingHorizontal": 24,
          },
          {},
        ],
      ]
    }
    testID="container-main-view"
  >
    <Text
      body700={true}
      h5={true}
      style={
        [
          {
            "backgroundColor": "transparent",
            "color": "#20303C",
            "writingDirection": "ltr",
          },
          undefined,
          undefined,
          undefined,
          undefined,
          {},
          {
            "textAlign": "center",
          },
          undefined,
          undefined,
          undefined,
        ]
      }
    >
      Error processing request
    </Text>
    <View
      style={
        [
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          [
            {
              "backgroundColor": "transparent",
              "height": 8,
              "width": 8,
            },
            undefined,
          ],
        ]
      }
      testID="padder-view"
    />
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#2E61DE",
          "borderRadius": 8,
          "marginRight": 8,
          "opacity": 1,
          "paddingHorizontal": 16,
          "paddingVertical": 8,
        }
      }
      testID="error-button"
    >
      <Text
        button={true}
        style={
          [
            {
              "backgroundColor": "transparent",
              "color": "#20303C",
              "writingDirection": "ltr",
            },
            undefined,
            {
              "color": "#FFFFFF",
            },
            undefined,
            undefined,
            {},
            {
              "textAlign": "center",
            },
            undefined,
            undefined,
            undefined,
          ]
        }
        white={true}
      >
        Try Again
      </Text>
    </View>
  </View>
</View>
`;
